### ISSUE TITLE
Bug: Deployments causing downtime - 503/504s

### ISSUE THREAD

**Issue by T-J-L** [aws/aws-app-mesh-roadmap#362]
**Summary**

Hello, I'm using App Mesh on ECS. The issue I'm facing is when performing a new deployment, requests are dropped for a significant amount of time. I believe I'm following the best practices [here](https://docs.aws.amazon.com/app-mesh/latest/userguide/best-practices.html), and already have health checks and retries configured. I suspect I must be missing something.

**Steps to Reproduce**

I can replicate the issue, by using the force new deployment option.

I setup a simple GET request which is hit every 2 seconds. The setup is `API GW > VPC Link > NLB > Service A > Service B`

I force new deployment of Service B, nothing else changes.

 The results were...

```
15:07:45 - (service B) has reached a steady state.
15:07:45 - (service B) (deployment B) deployment completed.

< Consistent 200s >

15:08:05 - First Failed Request

< 503 upstream connect error or disconnect/reset before headers. reset reason: connection failure >
< 504 upstream request timeout >

15:13:41 - Next 200 after consistent failures 

< Intermittent 200 / 503 / 504 >

15:16:01 - Back to consistent 200
```

As you can see this is 5 minutes of downtime, followed by 3 minutes of intermittent failures. The expected behaviour is no dropped requests. 

Any suggestions welcome.

**Comment by T-J-L** [comment]
Additional information: I'm also using Cloud Map for service discovery, with DNS & HTTP mode. DNS TTL's are all set to 30s and that's reflected in Route 53.

**Comment by Y0Username** [comment]
@T-J-L Are you still seeing this issue?
If yes, could you give us some more context on this issue? 
- How many tasks does each service have? 
- What kind of deployment are you doing?
- Are you using a virtual router between these services with a route that has retries enabled?

**Comment by T-J-L** [comment]
Yes I am, I'll get back to you with detail next week.

**Comment by Y0Username** [comment]
@T-J-L, following up on this.

**Comment by aruandre** [comment]
we are facing the same issue
Private DNS namepsace, Fargate with task Healthchecks and virtual router retries

**Comment by Y0Username** [comment]
@T-J-L sorry for the delay on this.
- Are you using CloudMap or DNS for service discovery?
- Can you check if this happens with 2 tasks and 150% deployment velocity?

@T-J-L  and @aruandre 
- Do you have active health checks enabled between service A and B?
- Can you check if the endpoints are getting updated in your Envoy configuration after deployment?
   You can check this by getting the Envoy config using command: `curl <task_ip>:9901/config_dump?include_eds`
   Get this config for Service A:
    - Before Service B deployment
    - Just after Service B deployment (when you see 5xx)
    - After Service B deployment (when you see 2xx)
   Basically, we're looking to see if Service A Envoy has the updated endpoint addresses for Service B after deployment.

**Comment by aruandre** [comment]
@T-J-L 

What do you mean by "active healthchecks between tasks"? We do have container healthchecks enabled if that is what you mean.

As for the config dump - I'm not seeing any difference in the endpoints block before, during and after the errors
[config_dump_post_5xx.log](https://github.com/aws/aws-app-mesh-roadmap/files/7562287/config_dump_post_5xx.log)
[config_dump_200.log](https://github.com/aws/aws-app-mesh-roadmap/files/7562288/config_dump_200.log)
[config_dump_5xx.log](https://github.com/aws/aws-app-mesh-roadmap/files/7562289/config_dump_5xx.log)
.

**Comment by T-J-L** [comment]
@Y0Username No problem. I'm using CloudMap for service discovery and all services have health checks. Will need some time to grab the logs.

**Comment by Y0Username** [comment]
@aruandre Envoys can health check the downstream service. For example, if you configure health checks under Service B Listener, Service A can health check Service B. [Here](https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_HealthCheckPolicy.html) is the AppMesh API reference.

Also, thanks for the configs. I have looked at them and I have a couple of questions:
*(I'm assuming your network path is `Service A --> Service B` and the configs you provided are from Service A.)*
- Do you also have 1 task running on Service B?
- EDS (Endpoint Discovery Service) configs will only update if you are using CloudMap for service discovery,
   in case of DNS, you should see cluster updating it's IPs in the Envoy logs if you set it to `debug` level.
- Assuming you have provided Service A configs, I don't see the cluster that is talking to Service B. I believe it's passing through `ALLOW_ALL` (`cds_egress_appmesh_mesh-allow-all`) cluster.
- Are you not explicitly providing Service B as a backend to Service A in the VirtualNode configuration?

**Comment by aruandre** [comment]
@Y0Username 

we had 2 tasks of service B running.
CloudMap for service discovery, sorry for the confusion.
service B is set as a backend for service A's VirtualNode.

I think we had VirtualNode healthchecks set for service B. I'll double check and run some tests once more.

**Comment by Y0Username** [comment]
@aruandre thanks for confirming.

From the envoy configs you provided for `Service A`, I do not see a cluster for `Service B`. If you have modeled `Service B` as a backend for `Service A`, there should be a cluster config. 
I also do not see EDS configs vended from AppMesh. If you're using CloudMap, there should be `Service B`'s cluster endpoints listed in the EDS config you provided.

Could you please double check these?

**Comment by aruandre** [comment]
@Y0Username 

we realized we had a misconfiguration in our virtual nodes' backends. after fixing this and applying virtual node healthchecks we were still receiving 5xx's.

we run a couple of test with tcpdump and observed that service A's envoy stats showed the backends correctly even when new tasks were deployed and removed. 
at the same time the old task which was to be removed was receiving traffic up until it was eventually shutdown. at the same time the new task started receiving traffic roughly 10 seconds after the old one was shutdown.

**Comment by Y0Username** [comment]
@aruandre thanks for the info.

Could you check if the deployment timeline aligns in your CloudMap namespace. For `Service B`, when you deploy, could you check your  CloudMap namespace for that service to see when the registration of new IP happens and when deregistration of old IP happens? And could you check if that aligns with EDS config on your `Service A` Envoy?

**Comment by aruandre** [comment]
@Y0Username 

the timeline in ECS's events
2021-11-22 10:21:15 service B has started 1 tasks
2021-11-22 10:23:08 service B registered 1 targets in target-group
2021-11-22 10:23:56 service B deregistered 1 targets in target-group
2021-11-22 10:23:56 service B has begun draining connections on 1 tasks
2021-11-22 10:26:05 service B has stopped 1 running tasks
2021-11-22 10:27:21 service B deployment completed
2021-11-22 10:27:21 service B has reached a steady state

timeline in CloudMap
2021-11-22 10:23:10 we see a new instance being registered into CloudMap for service B
2021-11-22 10:23:48 new service B instances becomes healthy, the "old" instance of service B has been healthy all along
2021-11-22 10:26:54 the "old" instance of service B is declared unhealthy
2021-11-22 10:27:07 the "old" instance is removed from CloudMap

timeline for service A envoy:9901/clusters
2021-11-22 10:23:51 envoy picks up the new instance - health_flags::healthy
2021-11-22 10:26:49 the "old" instance of service B is removed from envoy

until 2021-11-22 10:28:08 requests returned 200's only
2021-11-22 10:28:09 503's
2021-11-22 10:28:21 "old" task stopped receiving traffic and became unresponsive
2021-11-22 10:29:13 "new" task received its first actual traffic, before that it was only VirtualNode healthchecks
2021-11-22 10:29:24 200's again

so the times for CloudMap and service A envoy seem to add up.

why would the traffic still be routed to the old task while a new one is healthy and ready to receive it?

**Comment by abaptiste** [comment]
Hi @aruandre,

We are working on trying to replicate this issue.

In the meantime, if this is readily reproducible for you, could you enable `debug` logs on your Envoy nodes on `Service A` and force a deployment for `Service B`?  The debug logs will show the upstream connection activity to `Service B` and which addresses are being used.   The logs should also contain endpoint updates so that we can determine when `Service B` was updated with new hosts.

I see you are using CloudMap and also your DNS TTL's are 30s.   Envoy periodically refreshes DNS hosts every 5 second by default, so in either scenario minutes of downtime is abnormal.

**Comment by aruandre** [comment]
@abaptiste 

any way I can share the logs with you in private?
our TTL's are set to 60s

**Comment by Y0Username** [comment]
@aruandre I believe you have a ticket with us internally, you can either share it there or you can hop on [App Mesh  slack channel](https://join.slack.com/t/awsappmesh/shared_invite/zt-dwgbt85c-Sj_md92__quV8YADKfsQSA) and DM either of us. 

What surprises me is that  `10:26:49 the "old" instance of service B is removed from envoy` and still keeps forwarding traffic there until `10:28:09 503's`. Could you also send us the complete `Service A` Envoy config? (`curl <service_A_task_ip>:9901/config_dump?include_eds`).

When does the health checks for old instance starts failing? What are your health check config?

Also, you could check out [Outlier Detection](https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_Listener.html#appmesh-Type-Listener-outlierDetection) in AppMesh. Which is sort of passive health checking mechanism. [Here's an example](https://github.com/aws/aws-app-mesh-examples/tree/main/walkthroughs/howto-outlier-detection) of how outlier detection works.

In any case, even without outlier detection you should not be seeing minutes of outage.

**Comment by Y0Username** [comment]
For posterity, we were able to RCA @aruandre's issue. All their traffic was being routed through `ALLOW_ALL` [egress cluster](https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_EgressFilter.html), due to a mis-configuration in Virtual Service name. Virtual Service names must be FQDNs that are discoverable in your VPC. Since the mesh egress allow all cluster uses `ORIGINAL_DST` [service discovery type](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/service_discovery#original-destination), the endpoints that were being routed to were from the app's DNS cache.  

@T-J-L could you please confirm if you have `ALLOW_ALL` mesh egress filter configured and/or your Virtual Service names are FQDNs?

**Comment by T-J-L** [comment]
@Y0Username, thanks for the update. That does sound similar to my setup.

I do have the `ALLOW_ALL` egress filter configured.

The URLs I'm using between services are of the form `http://<service_name>.<cloudmap hosted zone>:<port>`.

What's the best way to confirm traffic is being routed correctly?

**Comment by Y0Username** [comment]
Could you confirm the name your Virtual Service that points to `Service B` and the backend name configured on your `Service A` Virtual Node?

You could enable debug logging for `Service A` Envoy, by setting `ENVOY_LOG_LEVEL` env to `debug` on the task. 
Now, if you make a request from `Service A` to `Service B`, the `Service A` Envoy should log which cluster it is using to reach `Service B`, if you see something like `cds_egress_appmesh_mesh-allow-all`, it is using the mesh allow all cluster, which is essentially a catch-all cluster, which does not have endpoint information for `Service B`. It just uses the IPs cached by the `Service A` App's DNS.

**Comment by T-J-L** [comment]
Found it:

`[2021-12-06  22:01:10.107][36][debug][router] [source/common/router/router.cc:457]  [C433][S11408447762775016216] cluster  'cds_egress_scoreboard-8801183_mesh-allow-all' match for URL`

**Comment by Y0Username** [comment]
@T-J-L, you seem to have the same problem. I bet all your traffic is being routed through `cds_egress_appmesh_mesh-allow-all` cluster. Your Virtual Service name should be FQDN, i.e `<service_name>.<cloudmap namespace>` and same goes for the backend names.

Refer this [doc](https://docs.aws.amazon.com/app-mesh/latest/userguide/virtual_services.html), for Virtual Service naming conventions. 

In your example: 
```
Service A
Virtual service name: score-62fac12.<cloudmap.hosted.zone>

Virtual node A
Virtual node name: score-62fac12
Service name: score-62fac12.<cloudmap.hosted.zone>
Backend: cloud-5e8fc99.<cloudmap.hosted.zone>

Service B
Virtual service name: cloud-5e8fc99.<cloudmap.hosted.zone>

Virtual node B
Virtual node name: cloud-5e8fc99
Service name: cloud-5e8fc99.<cloudmap.hosted.zone>
```

**Comment by T-J-L** [comment]
@Y0Username, you are correct, see my cross post above. I'll recreated my services and confirm.

**Comment by T-J-L** [comment]
@Y0Username, Thanks again for your help. I’m getting somewhere now as all requests are routed via the correct cluster:

`[C934][S10174538067616191874] cluster 'cds_egress_scoreboard-8801183_cloud-5e8fc99_http_8021' match for URL`

And my config now looks like:

```
Service A
Virtual service name: score-62fac12.cloudMap-526c587

Virtual node A
Virtual node name: score-62fac12
Service name: score-62fac12.cloudMap-526c587
Backend: cloud-5e8fc99.cloudMap-526c587

Service B
Virtual service name: cloud-5e8fc99.cloudMap-526c587

Virtual node B
Virtual node name: cloud-5e8fc99
Service name: cloud-5e8fc99.cloudMap-526c587
```

Unfortunately, I am now getting 503’s for the HTTP requests. 

My URL structure is: `http://score-62fac12.cloudMap-526c587:<port>/path`

The requests are being directed to the correct cluster as above. I have virtual routers setup for each service, with the “/” catch all prefix.
 
Looking through the logs I see the issue is `no healthy host for HTTP connection pool
`. Any suggestions on possible issues?

**Comment by T-J-L** [comment]
I can confirm everything is working as excpected. Thanks again for all your assistance.
